LOKI2R.GCL.R

```diff
@@ -52,132 +87,157 @@ LOKI2R.GCL=function(sillyvec,username,password){
+      ) %>%
+      dplyr::arrange(FK_FISH_ID)
+    
+    message(message0)
+    
+    assign(paste0(silly, ".gcl"), silly_df, pos = 1, .GlobalEnv)
+    
+    tibble::tibble(silly = silly, missing_loci = setdiff(loci, sillydata$LOCUS %>% unique()))
+    
+  }) %>% 
+    dplyr::bind_rows()# silly
+  
+  if(length(missing_sillys)>=1){
+    
+    message(paste0("The following sillys had no data in LOKI for the loci in LocusControl: ", paste0(missing_sillys, collapse = ", ")))
+    
     }
-      
-    scores <- array(data=NA,dim=c(nind,length(loci),max(ploidy)),dimnames=list(ids,loci,paste0("Dose",seq(max(ploidy)))))
+  
+  if(nrow(missing_loci)>=1){
     
-    sillyloci <- sort(unique(sillydata$LOCUS)) #These are the loci available for the current silly, this is needed to subset the tapply
+    n_missing <- missing_loci %>% 
+      dplyr::group_by(silly) %>% 
+      dplyr::summarize(n_miss = length(missing_loci) %>% as.character()) %>% 
+      tidyr::unite(col = silly_n_miss, silly, n_miss, sep = "(missing=") %>% 
+      dplyr::mutate(silly_n_miss = paste0(silly_n_miss, ")")) %>% 
+      dplyr::pull()
     
-    scores[ids,sillyloci,1] <- tapply(X=as.vector(sillydata$ALLELE_1),INDEX=list(sillydata$FISH_ID,sillydata$LOCUS),FUN=unique)[ids,sillyloci]
-      
-    scores[ids,sillyloci,2] <- tapply(X=as.vector(sillydata$ALLELE_2),INDEX=list(sillydata$FISH_ID,sillydata$LOCUS),FUN=unique)[ids,sillyloci]
-      
-    counts <- array(NA,c(nind,nloci,max(nalleles)),dimnames=list(ids,loci,paste("Allele ",1:max(nalleles),sep="")))
-      
-    for(ind in ids){
-
-      for(locus in loci){
-
-        for(al in seq(nalleles[locus])){
-
-	    counts[ind,locus,al] <- sum(scores[ind,locus,seq(ploidy[locus])]==alleles[[locus]][al])
-
-        }#al          
-       
-	}#locus
-
-      counts0=counts[ind,,]
-
-      counts0[is.na(counts0)] <- 0
-
-      zeroBOOL <- apply(counts0,1,sum)!=ploidy
-
-      counts[ind,zeroBOOL,] <- NA
-           
-    }#ind
-      
-    attributes <- sillydata[,attnames]
-      
-    attributes <- aggregate(attributes[,-match("FISH_ID",attnames)],list(FISH_ID=attributes$FISH_ID),unique)
-      
-    attributes <-  data.frame(attributes,SillySource=sillyvials,row.names=ids,stringsAsFactors = FALSE)
-
-    names(attributes)[names(attributes)=="FISH_ID"] <- "FK_FISH_ID"   
+    message(paste0("The following sillys were missing data for one or more loci: ", paste0(n_missing, collapse = ", ")))
     
-    attributes$CAPTURE_DATE <- as.POSIXct(attributes$CAPTURE_DATE)
+    message(paste0("A table of loci missing data for each *.gcl object can be found here: ", paste0(getwd(), "/LOKI2R_missing_loci.csv")))
     
-    attributes$END_CAPTURE_DATE <- as.POSIXct(attributes$END_CAPTURE_DATE)
-      
-    message(message0)
-      
-    assign(paste(silly,".gcl",sep=""),list(counts=counts,scores=scores,n=nind,attributes=attributes),pos=1)
-
-  }#silly 
-
+    readr::write_csv(missing_loci, path = "LOKI2R_missing_loci.csv")

```

Comment:
I like the idea of returning missing loci to the environment, but I guess I was leaning more towards assigning it to an object in your workspace, rather than writing out a file to your working directory.